modules/tftp/http_downloads.rb

```diff
@@ -0,0 +1,318 @@
+
+      def initialize(now = Time.now)
+        @m = Mutex.new
+        @file_length = 0
+        @downloaded = 0
+        @timestamp = now
+        @stopped = false
+      end
+
+      def file_length
+        @m.synchronize { @file_length }
+      end
+
+      def downloaded
+        @m.synchronize { @downloaded }
+      end
+
+      def file_length=(length)
+        @m.synchronize { @file_length = length }
+      end
+
+      def update_downloaded(chunk_length)
+        @m.synchronize { @downloaded += chunk_length }
+      end
+
+      def progress
+        (fl = file_length) == 0 ? 0 : ((downloaded.to_f/fl) * 100).round(2).to_i
+      end
+
+      def reset_error
+        @last_error = nil
+      end
+
+      def stopped?
+        @stopped
+      end
+
+      def to_hash
+        {:file_length => file_length, :downloaded => downloaded, :progress => progress, :timestamp => timestamp, :last_error => last_error}
+      end
+    end
+
+    class Downloader
+      ATTEMPT_DELAY = 10
+
+      def initialize(parsed_url, dst_path, status)
+        @stop = false
+        @status = status
+        @parsed_url = parsed_url
+        @dst_path = dst_path
+      end
+
+      def size
+        Net::HTTP.start(@parsed_url.host, @parsed_url.port, :use_ssl => @parsed_url.scheme == 'https') do |http|
+          head = Net::HTTP::Head.new(@parsed_url)
+          http.request(head) do |response|
+            raise ::Proxy::TFTP::HttpError.new(response.code, response.message) unless response.is_a?(Net::HTTPSuccess)
+            response['content-length'].to_i rescue 0
+          end
+        end
+      end
+
+      def start(start_position = 0)
+        Net::HTTP.start(@parsed_url.host, @parsed_url.port, :use_ssl => @parsed_url.scheme == 'https') do |http|
+          get = Net::HTTP::Get.new(@parsed_url, (range_header(start_position) if start_position != 0))
+          http.request(get) do |response|
+            raise ::Proxy::TFTP::HttpError.new(response.code, response.message) unless response.is_a?(Net::HTTPSuccess)
+            @status.file_length = response['content-length'].to_i rescue 0
+            @status.update_downloaded(start_position) if response.code == '206'
+            open(@dst_path, (response.code == '206' ? 'ab' : 'wb')) do |io|
+              if @stop
+                @status.stopped = true
+                break
+              end
+              response.read_body {|b| @status.update_downloaded(b.length); io.write(b)}
+            end
+          end
+        end
+      end
+
+      def restart
+        start(File.size?(@dst_path) || 0)
+      end
+
+      def range_header(range_start)
+        {'Range' => "bytes=#{range_start}-"}
+      end
+
+      def stop
+        @stop = true
+      end
+    end
+
+    attr_reader :status, :m
+
+    def initialize(dst_path, url, status = Status.new)
+      @status = status
+      @dst = dst_path
+      @url = URI.parse(url)
+      @downloader = nil
+      @m = Mutex.new
```

Comment:
Add descriptive name to all mutexes in this patch with comments what is the purpose. While this one is easier to understand, the one below is not that clear. I would like to prevent future changes mis-use of mutexes which is a typical hard to debug error.